### PR TITLE
#4173. Add tests for NULL types

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+/// @description Check that UP(`T1`, `T2`) = `T2` if `T1 != T2`, `T1` and `T2`
+/// are both not a TOP, BOTTOM or intersection types and NULL(`T1`), NULL(`T2`)
+/// and MOREBOTTOM(`T1`, `T2`).
+/// @author sgrekhov22@gmail.com
+/// @issue 63908
+
+// ignore_for_file: unused_local_variable
+
+void f1(Never n) {
+  var v = (1 > 2) ? n : null;
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A10_t01.dart]: `NULL(Never)` doesn't hold. The types that satisfy `NULL` are `Null`, `Null?`, and `T?` where `BOTTOM(T)` (that is, `Never?` and `X?` where `X extends Never`; `(X & B)?` where `BOTTOM(B)` doesn't count because no types of the form `(X & B)?` exist).

So we'd basically have a matrix with those entries on both axes, and then we'd filter out the combinations which have already been handled in earlier rules. In particular, the main diagonal is gone because it is handled by `UP(T, T)`.

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+/// @assertion We define the upper bound of two types `T1` and `T2` to be
+/// UP(`T1`,`T2`) as follows.
+/// ...
+/// - UP(`T1`, `T2`) where NULL(`T1`) and NULL(`T2`) =
+///   - `T2` if MOREBOTTOM(`T1`, `T2`)
+///   - `T1` otherwise
+///
+/// @description Check that UP(`T1`, `T2`) = `T2` if `T1 != T2`, `T1` and `T2`
+/// are both not a TOP, BOTTOM or intersection types and NULL(`T1`), NULL(`T2`)
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A10_t01.dart]: I think it's easier to see that it is a list of cases if only the last element has 'and'.
```suggestion
/// are both not a TOP, BOTTOM, or intersection type, if NULL(`T1`), NULL(`T2`),
```
We could add 'if' before every element in the list, but I think it's OK to have it only at the spot where it gets hard to remember that we're "iffing" every element in the list. ;-)

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+/// @author sgrekhov22@gmail.com
+/// @issue 63908
+
+// ignore_for_file: unused_local_variable
+
+void f1(Never n) {
+  var v = (1 > 2) ? n : null;
+  // UP(Never, Null) = Null, because MOREBOTTOM(Never, T) = true
+  // Let's check that `v` is not assignable to a non-nullable variable
+  int i = v;
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A10_t01.dart]: Wouldn't it work to test directly that `v.expectStaticType<Exactly<Null>>`?

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+/// ...
+/// - UP(`T1`, `T2`) where NULL(`T1`) and NULL(`T2`) =
+///   - `T2` if MOREBOTTOM(`T1`, `T2`)
+///   - `T1` otherwise
+///
+/// @description Check that UP(`T1`, `T2`) = `T1` if `T1 != T2`, `T1` and `T2`
+/// are both not a TOP, BOTTOM or intersection types and NULL(`T1`), NULL(`T2`)
+/// and not MOREBOTTOM(`T1`, `T2`).
+/// @author sgrekhov22@gmail.com
+/// @issue 63908
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A10_t02.dart:16]: Similar to A10_ t01.

File: TypeSystem/upper-lower-bounds/up_A11_t01.dart [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion We define the upper bound of two types `T1` and `T2` to be
+/// UP(`T1`,`T2`) as follows.
+/// ...
+/// - UP(`T1`, `T2`) where NULL(`T1`) =
+///   - `T2` if `T2` is nullable
+///   - `T2*` if `Null <: T2` or `T1 <: Object` (that is, `T1` or `T2` is
+///     legacy)
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A11_t01.dart]: I'd recommend that we simply remove every rule or subrule which is dealing with legacy types. They do not exist in relevant versions of Dart.
```suggestion
```

File: TypeSystem/upper-lower-bounds/up_A11_t01.dart [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+///   - `T2?` otherwise
+///
+/// @description Check that UP(`T1`, `T2`) = `T2` if `T1 != T2`, `T1` and `T2`
+/// are both not a TOP, BOTTOM or intersection types, NULL(`T1`) is `true`, not
+/// NULL(`T2`) and `T2` is nullable.
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A11_t01.dart]:
```suggestion
/// @description Check that UP(`T1`, `T2`) = `T2` if `T1 != T2`, `T1` and `T2`
/// are both not a TOP, BOTTOM or intersection type, if NULL(`T1`), not
/// NULL(`T2`), and `T2` is nullable.
```

File: TypeSystem/upper-lower-bounds/up_A11_t01.dart [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+import 'dart:async';
+import '../../Utils/static_type_helper.dart';
+import 'up_lib.dart';
+
+void f1(num? n) {
+  var v = (1 > 2) ? null : n; // UP(Null, num?) = num?
+  v.expectStaticType<Exactly<num?>>();
+}
+
+void f2(Never? t1, num? n) {
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A11_t01.dart:26]: OK, so we're just checking the case NULL(`T1`) using the types `Null` and `Never?`. That's fine, given that the two A10 tests have all cases.

File: TypeSystem/upper-lower-bounds/up_A11_t01.dart (hunk 6 above)
Comment (eernstg) [reply TypeSystem/upper-lower-bounds/up_A11_t01.dart]: A similar comment would apply to the `@description` of more tests below.

**Review by eernstg — COMMENTED** [review]
Looks good! There is one test where a type should be NULL but isn't, so that one needs an adjustment. There's another comment on the descriptions.

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart (hunk 2 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A10_t01.dart]: Updated.

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart (hunk 1 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A10_t01.dart]: Ah yes, indeed! Replaced by `X?` where `X extends Never` and moved to the test where we are checking not MOREBUTTOM(because MOREBOTTOM(X?, Null) == false).

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart (hunk 3 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A10_t01.dart]: Yes, `v.expectStaticType<Exactly<X?>>();`.

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart (hunk 4 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A10_t02.dart:16]: Updated.

File: TypeSystem/upper-lower-bounds/up_A11_t01.dart (hunk 5 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A11_t01.dart]: Removed here and below.

File: TypeSystem/upper-lower-bounds/up_A11_t01.dart (hunk 6 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A11_t01.dart]: Updated here and below.

**Review by sgrekhov — COMMENTED** [review]
Thank you. Updated. PTAL.

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+void f2(Null? n) { // ignore: unnecessary_question_mark
+  // UP(Null, Null?) = dynamic for historical reasons though MOREBOTTOM(Null, Null?) = true
+  var v = (1 > 2) ? null : n;
+  if (1 > 2) {
+    v.checkDynamic;
+  }
+}
+
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A10_t01.dart:48]: OK, so we have `(Null, Never?)`, `(Null, Null?)`, but not `(Null, X?)` where `X extends Never`, and not the opposite direction. But as I understand it, they're covered in A10_t02.

Still, I can't find `(Null, X?)`. Also I can't find pairs that don't include `Null` itself (like `(Null?, Never?)` or `(Null?, X?)` or `(Null?, Never?)`). Are they somewhere else?

I know, it's extremely nitpickish, but I don't think we have any other test that covers `NULL` and `UP` in those cases.

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart (hunk 4 above)
Comment (eernstg) [reply TypeSystem/upper-lower-bounds/up_A10_t02.dart:16]: Thanks!

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+import '../../Utils/static_type_helper.dart';
+
+// ignore_for_file: unused_local_variable
+
+void f1<X extends Never>(X? n) {
+  var v = (1 > 2) ? n : null;
+  // UP(X?, Null) = X?, because MOREBOTTOM(X?, Null) == false
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A10_t02.dart]: If `MOREBOTTOM(X?, Null)` is false, wouldn't that yield `UP(X?, Null) = Null`?

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+  // Let's check that `v` is not assignable to a non-nullable variable
+  v.expectStaticType<Exactly<X?>>();
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A10_t02.dart]: We really can't see the difference, but I think it should be `Null`.

File: TypeSystem/upper-lower-bounds/up_A11_t01.dart (hunk 5 above)
Comment (eernstg) [reply TypeSystem/upper-lower-bounds/up_A11_t01.dart]: Thanks!

**Review by eernstg — COMMENTED** [review]
I'm afraid we still have a couple of things here.

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart (hunk 8 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A10_t01.dart:48]: Added, thank you.

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart (hunk 9 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A10_t02.dart]: UP(`T1`, `T2`) = `T1` if MOREBOTTOM(X?, Null) == false. `X?` in this case.

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart (hunk 10 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A10_t02.dart]: Please see my comment above.

**Review by sgrekhov — COMMENTED** [review]
Updated, the missing cases added. PTAL

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  var v2 = (1 > 2) ? v : 42 as Object?;
+  v2.proofNotDynamic;
+//   ^^^^^^^^^^^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
+}
+
+void f2(Null? n) {
+  // UP(Null, Null?) = dynamic for historical reasons though MOREBOTTOM(Null, Null?) = true
+  var v = (1 > 2) ? null : n;
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A10_t01.dart]: It's true that the inferred type of `v` is `dynamic`, but the type of the initializing expression is still `Null`, so we should be able to test that directly:

```dart
  ((1 > 2) ? null : n).expectStaticType<Exactly<Null>>;
```
With that, we shouldn't need lines 42-44.

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart (hunk 8 above)
Comment (eernstg) [reply TypeSystem/upper-lower-bounds/up_A10_t01.dart:48]: Thanks!

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+
+import '../../Utils/static_type_helper.dart';
+
+// ignore_for_file: unused_local_variable, unnecessary_question_mark
+
+void f1(Never? n) {
+  var v = (1 > 2) ? null : n;
+  // UP(Null, Never?) = Never?, because MOREBOTTOM(Null, T) = true
+  // It's not possible to check that `v` is `Never?`. Let's check that it is
+  // assignable to both `String` and `int` and is not `dynamic`
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A10_t01.dart]: But `Never?` is not assignable to `int` and `String`, we'd need to strip off the `?` part first.

For example, before line 27: `if (v == null) return-or-fail;`. When `v` has been promoted to a non-null type we can assign to `int` and `String`.

If `v` has the type `dynamic` then it still works because the `if ...` lets `v` keep the type `dynamic`, so we will still detect that case and fail as we should.

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+import '../../Utils/static_type_helper.dart';
+
+// ignore_for_file: unused_local_variable, unnecessary_question_mark
+
+void f1(Never? n) {
+  var v = (1 > 2) ? n : null;
+  // UP(Never?, Null) = Never?, because MOREBOTTOM(T, Null) = false
+  // It's not possible to check that `v` is `Never?`. Let's check that it is
+  // assignable to both `String` and `int` and is not `dynamic`
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A10_t02.dart:26]: Needs to return/fail if `v == null`, as in the previous test.

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+  var v2 = (1 > 2) ? v : 42 as Object?;
+  v2.proofNotDynamic;
+//   ^^^^^^^^^^^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
+}
+
+void f2(Null? n) {
+  // UP(Null?, Null) = dynamic for historical reasons though MOREBOTTOM(Null?, Null) = false
+  var v = (1 > 2) ? n : null;
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A10_t02.dart]: We can test the type of the expression `(1 > 2) ? n : null` directly.

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart (hunk 9 above)
Comment (eernstg) [reply TypeSystem/upper-lower-bounds/up_A10_t02.dart]: Ah, of course!

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart (hunk 10 above)
Comment (eernstg) [reply TypeSystem/upper-lower-bounds/up_A10_t02.dart]: My bad again.

**Review by eernstg — COMMENTED** [review]
Very nice! Just a couple of comments. In particular, the case where the variable type is `dynamic` because the initializing expression has static type `Null` can be tested directly. Another test seems to expect that `Never?` can be assigned to `int` and `String`, but that should be rejected as a compile-time error.

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart (hunk 11 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A10_t01.dart]: How interesting. Thanks! Updated.

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart (hunk 12 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A10_t01.dart]: Added. The actual type of `v` here is `dynamic`. It's sdk#63908

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart (hunk 13 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A10_t02.dart:26]: Added.

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart (hunk 14 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A10_t02.dart]: Updated.

**Review by sgrekhov — COMMENTED** [review]
Thanks. Updated. PTAL.

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart (hunk 11 above)
Comment (eernstg) [reply TypeSystem/upper-lower-bounds/up_A10_t01.dart]: Thanks!

File: TypeSystem/upper-lower-bounds/up_A10_t01.dart (hunk 12 above)
Comment (eernstg) [reply TypeSystem/upper-lower-bounds/up_A10_t01.dart]: Very good, thanks!

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart [hunk 15]
```diff
@@ -37,11 +40,10 @@ void f1(Never? n) {
...
 void f2(Null? n) {
-  // UP(Null?, Null) = dynamic for historical reasons though MOREBOTTOM(Null?, Null) = false
-  var v = (1 > 2) ? n : null;
-  if (1 > 2) {
-    v.checkDynamic;
-  }
+  // In the case var v = (1 > 2) ? n : null; the inferred type of `v` is
+  // `dynamic` for historical reasons. Though MOREBOTTOM(Null?, Null) = false
+  // and the expression should has type `Null`.
+  ((1 > 2) ? n : null).expectStaticType<Exactly<Null>>();
```
Comment (eernstg) [inline TypeSystem/upper-lower-bounds/up_A10_t02.dart:46]: I think we should ensure that the type is not `dynamic` as well:

```dart
  ((1 > 2) ? n : null).proofNotDynamic;
//                     ^^^^^^^^^^^^^^^ and so on.
```

**Review by eernstg — COMMENTED** [review]
Just one location needs to check that a type isn't dynamic!

File: TypeSystem/upper-lower-bounds/up_A10_t02.dart (hunk 15 above)
Comment (sgrekhov) [reply TypeSystem/upper-lower-bounds/up_A10_t02.dart:46]: Updated.

**Review by eernstg — APPROVED** [review]
LGTM!
